finetrainers/trainer.py

```diff
@@ -479,30 +514,39 @@ def prepare_optimizer(self) -> None:
         params_to_optimize = [transformer_parameters_with_lr]
         self.state.num_trainable_parameters = sum(p.numel() for p in transformer_lora_parameters)
 
-        # TODO(aryan): add deepspeed support
+        use_deepspeed_opt = (
+            self.state.accelerator.state.deepspeed_plugin is not None
+            and "optimizer" in self.state.accelerator.state.deepspeed_plugin.deepspeed_config
+        )
         optimizer = get_optimizer(
             params_to_optimize=params_to_optimize,
             optimizer_name=self.args.optimizer,
-            learning_rate=self.args.lr,
+            learning_rate=self.state.learning_rate,
             beta1=self.args.beta1,
             beta2=self.args.beta2,
             beta3=self.args.beta3,
             epsilon=self.args.epsilon,
             weight_decay=self.args.weight_decay,
+            use_deepspeed=use_deepspeed_opt,
         )
 
         num_update_steps_per_epoch = math.ceil(len(self.dataloader) / self.args.gradient_accumulation_steps)
         if self.state.train_steps is None:
             self.state.train_steps = self.state.train_epochs * num_update_steps_per_epoch
             self.state.overwrote_max_train_steps = True
 
+        use_deepspeed_lr_schd = (
+            self.state.accelerator.state.deepspeed_plugin is not None
+            and "scheduler" in self.state.accelerator.state.deepspeed_plugin.deepspeed_config
+        )
         lr_scheduler = get_scheduler(
             name=self.args.lr_scheduler,
             optimizer=optimizer,
             num_warmup_steps=self.args.lr_warmup_steps * self.state.accelerator.num_processes,
             num_training_steps=self.state.train_steps * self.state.accelerator.num_processes,
             num_cycles=self.args.lr_num_cycles,
             power=self.args.lr_power,
+            use_deepspeed=use_deepspeed_lr_schd,
```

Comment:
The `get_scheduler` method comes from Diffusers, right? I don't think we have a `use_deepspeed` parameter, no? This errors out at the moment

Reply:
In my local copy I had modified that but forgot to push those changes. https://github.com/a-r-r-o-w/finetrainers/pull/139/commits/0c0a0a5e4abbe826f31167fe91423e7ef2580cb5 should have handled it rightly.

Reply:
Thanks!